api/types/database.go

```diff
@@ -489,7 +489,7 @@ func (d *DatabaseV3) setStaticFields() {
 func (d *DatabaseV3) CheckAndSetDefaults() error {
 	d.setStaticFields()
 	if err := d.Metadata.CheckAndSetDefaults(); err != nil {
-		return trace.Wrap(err)
+		return trace.Wrap(err, "database %q", d.GetName())
```

Comment:
Formatting of user messages depends on how the error is logged and what format the logger is using.

I went into my cluster's sql backend and modified a dynamic database manually to produce this error, because I'm actually not sure how it can be created in the backend otherwise.

We add an additional user message "failed to fetch resource: %q" here: https://github.com/gravitational/teleport/blob/517f93e1b42e76b400f796cef58ba9bd23858578/lib/cache/cache.go#L1287-L1290

and we log the error using `WithError` here: https://github.com/gravitational/teleport/blob/517f93e1b42e76b400f796cef58ba9bd23858578/lib/cache/cache.go#L849-L851

JSON logging example:
```
{
  "caller": "cache/cache.go:850",
  "component": "auth:1:cache",
  "error": "failed to fetch resource: \"db\"\n\tdatabase \"roach\"\n\t\tinvalid origin value \"bad-config\", must be one of [defaults config-file dynamic cloud kubernetes]",
  "level": "warning",
  "message": "Re-init the cache on error",
  "timestamp": "2023-01-31T11:48:08-08:00"
}
```

text example:
```
2023-01-31T11:52:46-08:00 WARN [AUTH:1:CA] Re-init the cache on error error:[
ERROR REPORT:
Original Error: *trace.BadParameterError invalid origin value &#34;bad-config&#34;, must be one of [defaults config-file dynamic cloud kubernetes]
Stack Trace:
	github.com/gravitational/teleport/api@v0.0.0/types/resource.go:413 github.com/gravitational/teleport/api/types.(*Metadata).CheckAndSetDefaults
	github.com/gravitational/teleport/api@v0.0.0/types/database.go:491 github.com/gravitational/teleport/api/types.(*DatabaseV3).CheckAndSetDefaults
	github.com/gravitational/teleport/lib/services/database.go:123 github.com/gravitational/teleport/lib/services.UnmarshalDatabase
	github.com/gravitational/teleport/lib/services/local/databases.go:48 github.com/gravitational/teleport/lib/services/local.(*DatabaseService).GetDatabases
	github.com/gravitational/teleport/lib/cache/collections.go:1465 github.com/gravitational/teleport/lib/cache.(*database).fetch
	github.com/gravitational/teleport/lib/cache/cache.go:1287 github.com/gravitational/teleport/lib/cache.(*Cache).fetch.func2
	golang.org/x/sync@v0.1.0/errgroup/errgroup.go:75 golang.org/x/sync/errgroup.(*Group).Go.func1
	runtime/asm_arm64.s:1172 runtime.goexit
User Message: failed to fetch resource: &#34;db&#34;
	database &#34;roach&#34;
		invalid origin value &#34;bad-config&#34;, must be one of [defaults config-file dynamic cloud kubernetes]] cache/cache.go:850
```

I think it's appropriate to add a user message that includes the name of the database so it's clear which one is the problem. We just need to fix our log formatting to handle user messages better (among other problems). There's an RFD for that here: https://github.com/gravitational/teleport/blob/428eb9eff08478e1758da2ce90aac99df079b001/rfd/0092-error-message-guidelines.md

Reply:
Definitely better with the database name included. Do you know if we have plan to change the UserMessage format?

I was curious about this only because I am not a big fan of having new lines in the message. I would probably prefer embed the name in the same message or wrap the name in the caller like:
https://github.com/gravitational/teleport/blob/26a623b5be96a8766541fabf01a232b3e57f1a82/lib/config/configuration.go#L1373-L1374

Reply:
> Do you know if we have plan to change the UserMessage format?
I think that will happen after the RFD is approved. I have a comment here: https://github.com/gravitational/teleport/pull/17305#discussion_r1073094451

For text formatting, I would like to get rid of the HTML escaping. There's a setting for that, I'm just not sure about the full implications of disabling that escaping, but it's part of the RFD discussion.

For JSON formatting, I would like us to modify the JSON formatter such that trace errors are marshaled without all the escaping quotes, newlines, tabs, etc. I tested a patch that does that, it's actually a pretty simple change. Here's what it could look like in server logs with JSON formatting:

```
{
  "caller": "cache/cache.go:805",
  "component": "auth:1:cache",
  "error": {
    "message": "invalid origin value `bad-config`, must be one of [defaults config-file dynamic cloud kubernetes]",
    "user_messages": [
      "database `pg`",
      "failed to fetch resource: `db`"
    ]
  },
  "level": "warning",
  "message": "Cache `auth` first init failed, continuing re-init attempts in background.",
  "timestamp": "2023-01-31T14:26:52-08:00"
}
```

Aside from that though, why wrap the name in the caller?

Reply:
> why wrap the name in the caller?

It's easier to wrap there as it's a single place you have to add the name, and you can add some context too e.g. `trace.Wrap(err, "Failed to validate database resource %q",  database.Name)`. Just personal preference.

>```
>"user_messages": [
>      "database `pg`",
>      "failed to fetch resource: `db`"
>]
>```
I know the RFD is still under review, but the `database 'pg'` feels more like a field than a message. Might be awesome if we can do something like this in this case `trace.Wrap(err).WithField("db_name", db.GetName())` and the field logger can add the extra fields from the trace.

Again, I started this comment as I was curious about the new line. I don't have a strong opinion on the wrapping.

@smallinsky what do you think?

Reply:
> It's easier to wrap there as it's a single place you have to add the name, and you can add some context too e.g. trace.Wrap(err, "Failed to validate database resource %q",  database.Name). Just personal preference.

It also looks more like a user message than a field, so this is probably the better way to go.

> we can do something like this in this case trace.Wrap(err).WithField("db_name", db.GetName()) and the field logger can add the extra fields from the trace.

There's a few places we already use `trace.WithField` and those fields are printed in debug mode with the text log formatter already as well. I'm in favor of adding those fields in the json formatter too, but i'm not sure exactly how we should handle that for user-facing logs in the CLI

Reply:
A don't have strong opinion either but personally i would incline to the `trace.WithField` approach where the json and text formatter print output looks readable. We can do something like: `"log.WithFields(trace.Wrap(err).GetFields()).WithError(err).Warn("...")"`

Reply:
I think this thread is a problem with the way we format user messages from `trace.Wrap`, but that should be addressed as a separate issue specifically for how we format errors and logs.

If I use
```
 if err := db.CheckAndSetDefaults(); err != nil { 
 	return trace.Wrap(err, "failed to validate database %v resource", db.Name) 
```

Then the database name will be repeated in cases where we have errors that already embed the database name like `trace.BadParameter("database %q protocol is empty", d.GetName())`, and every log will include the newline/tabs with the way we format logs/errors right now.
Additionally, I'd have to find all callers of this function and also wrap the database name there to consistently provide the db name in the error message.

How about for now I'll just remove changes that added the wrapped user message to avoid inserting newlines in json error logs. I'd rather not block the PR on this issue